include/pmacc/particles/algorithm/ForEach.hpp

```diff
@@ -22,157 +22,447 @@
-
-                                        bool const isPar = parIdx < numPartcilesInSupercell;
-                                        if(isPar)
-                                            accFunctor(acc, particle);
-                                    });
-
-                                frame = pb.getNextFrame(frame);
-                            }
-                        }
-                    };
+            using ForEachParticleInFrame
+                = lockstep::ForEach<lockstep::Config<frameSize * numFramesToSkip, T_numWorkers, 1>>;
 
-                } // namespace detail
-            } // namespace acc
+            DataSpace<dim> const m_superCellIdx;
+            ForEachParticleInFrame const forEachParticleInFrame;
+            T_ParBox m_particlesBox;
 
-            /** Run a unary functor for each particle of a species in the given area
+        public:
+            /** Provide the lockstep foreach executor to execute a functor for each particle within a set of frames.
              *
-             * Has a version for a fixed area, and for a user-provided mapper factory.
-             * They differ only in how the area is defined.
+             * Depending of the number of worker used it could be that more than one frame executed by all workers in
+             * parallel. The frame functor interface is therefore a context with one frame per virtual worker.
+             * It is guaranteed that the executor fits to the frameCtx provided by the frame functor interface. @see
+             * detail::FrameFunctorInterface
              *
-             * @warning Does NOT fill gaps automatically! If the
-             *          operation deactivates particles or creates "gaps" in any
-             *          other way, CallFillAllGaps needs to be called for the
-             *          species manually afterwards!
-             *
-             * @tparam T_Species type of the species
-             * @tparam T_Functor unary particle functor type which follows the interface of
-             *                   pmacc::functor::Interface<F, 1u, void>
-             *
-             * @param species species to operate on
-             * @param functor operation which is applied to each particle of the species
+             * @return Lockstep foreach executor to operate on frames.
+             */
+            DINLINE auto lockstepForEach() const
+            {
+                return forEachParticleInFrame;
+            }
+
+            /** Construct algoritm to operate on particles in a superCell
              *
-             * @{
+             * @param workerIdx workerIdx index of the worker: range [0;workerSize)
+             * @param particlesBox particles memory
+             *                     It is not allowed concurrently to add or remove particles during the
+             *                     execution of this algorithm.
+             * @param superCellIdx index of the superCell where particles should be processed
```

Comment:
Is it really a supercell index, or an index inside the given particle box?

Reply:
Perhaps a related thought - do we need to operate as the whole box and supercell index here? It would appear more logical to have a single supercell already as our collection of particles to operate on.

A user interface could still be a box + index if that is more convenient, a transition could also be made inside

Reply:
It is supercell index we need to operate on.

>  do we need to operate as the whole box and supercell index here? 

I also thought first of it but because of how pmacc is structured internally working with a supercell will not be easily possible without other refactorings in pmacc. There is no link from a supercell to the particle box you need to iterate over frames, therefore I decided to work with the supercell index.